File: src/executor/executor.cpp
#include "executor.h"
#include <iostream>
#include <algorithm>

Executor::Executor(StorageEngine& storage) : storage(storage) {}

void Executor::execute(const std::shared_ptr<Query>& query) {
    if (!query) return;

    switch (query->type) {
        case QueryType::CREATE_TABLE:
            execute_create_table(static_cast<const CreateTableQuery*>(query.get()));
            break;
        case QueryType::INSERT:
            execute_insert(static_cast<const InsertQuery*>(query.get()));
            break;
        case QueryType::SELECT:
            execute_select(static_cast<const SelectQuery*>(query.get()));
            break;
        default:
            std::cerr << "Unknown query type." << std::endl;
    }
}

void Executor::execute_create_table(const CreateTableQuery* query) {
    if (storage.create_table(query->table_name, query->columns)) {
        std::cout << "Table '" << query->table_name << "' created successfully." << std::endl;
    }
}

void Executor::execute_insert(const InsertQuery* query) {
    if (storage.insert_row(query->table_name, query->values)) {
        std::cout << "1 row inserted into '" << query->table_name << "'." << std::endl;
    }
}

bool Executor::evaluate_condition(const std::string& row_val, const std::string& op, const std::string& cond_val) {
    // Basic type inference: if both strings represent numbers, compare numerically
    try {
        int r_val = std::stoi(row_val);
        int c_val = std::stoi(cond_val);
        if (op == "=") return r_val == c_val;
        if (op == "<") return r_val < c_val;
        if (op == ">") return r_val > c_val;
    } catch (...) {
        // Fallback to string comparison
        if (op == "=") return row_val == cond_val;
        if (op == "<") return row_val < cond_val;
        if (op == ">") return row_val > cond_val;
    }
    return false;
}

void Executor::execute_select(const SelectQuery* query) {
    std::vector<std::string> columns;
    std::vector<std::vector<std::string>> rows;

    if (!storage.read_table(query->table_name, columns, rows)) {
        std::cerr << "Failed to read table '" << query->table_name << "'." << std::endl;
        return;
    }

    std::vector<int> col_indices;
    if (query->columns.size() == 1 && query->columns[0] == "*") {
        for (size_t i = 0; i < columns.size(); ++i) col_indices.push_back(i);
    } else {
        for (const auto& col_name : query->columns) {
            auto it = std::find(columns.begin(), columns.end(), col_name);
            if (it != columns.end()) {
                col_indices.push_back(std::distance(columns.begin(), it));
            } else {
                std::cerr << "Column '" << col_name << "' not found." << std::endl;
                return;
            }
        }
    }

    int where_col_idx = -1;
    if (query->has_where) {
        auto it = std::find(columns.begin(), columns.end(), query->where_clause.column);
        if (it != columns.end()) {
            where_col_idx = std::distance(columns.begin(), it);
        } else {
            std::cerr << "WHERE column '" << query->where_clause.column << "' not found." << std::endl;
            return;
        }
    }

    // Print headers
    for (size_t i = 0; i < col_indices.size(); ++i) {
        std::cout << columns[col_indices[i]];
        if (i < col_indices.size() - 1) std::cout << " | ";
    }
    std::cout << "\n";
    std::cout << std::string(col_indices.size() * 10, '-') << "\n";

    // Print rows
    int match_count = 0;
    for (const auto& row : rows) {
        if (query->has_where && where_col_idx != -1) {
            if (!evaluate_condition(row[where_col_idx], query->where_clause.op, query->where_clause.value)) {
                continue; 
            }
        }

        for (size_t i = 0; i < col_indices.size(); ++i) {
            std::cout << row[col_indices[i]];
            if (i < col_indices.size() - 1) std::cout << " | ";
        }
        std::cout << "\n";
        match_count++;
    }
    std::cout << "(" << match_count << " rows)\n";
}

File: src/main/java/com/coresql/engine/Executor.java
package com.coresql.engine;

import com.coresql.ast.*;

import java.util.ArrayList;
import java.util.List;

public class Executor {
    private final StorageEngine storage;

    public Executor(StorageEngine storage) {
        this.storage = storage;
    }

    public String execute(Query query) {
        if (query == null) return "Error: Query is null";

        switch (query.getType()) {
            case CREATE_TABLE -> { return executeCreateTable((CreateTableQuery) query); }
            case INSERT -> { return executeInsert((InsertQuery) query); }
            case SELECT -> { return executeSelect((SelectQuery) query); }
            default -> { return "Error: Unknown query type."; }
        }
    }

    private String executeCreateTable(CreateTableQuery query) {
        if (storage.createTable(query.tableName, query.columns)) {
            return "Table '" + query.tableName + "' created successfully.";
        }
        return "Error: Failed to create table '" + query.tableName + "'.";
    }

    private String executeInsert(InsertQuery query) {
        List<ColumnDefinition> schema = storage.readSchema(query.tableName);
        if (schema == null) {
            return "Error: Table '" + query.tableName + "' does not exist.";
        }

        if (schema.size() != query.values.size()) {
            return "Error: Column count mismatch. Expected " + schema.size() + ", got " + query.values.size();
        }

        // Validate types
        for (int i = 0; i < schema.size(); i++) {
            String expectedType = schema.get(i).type;
            String val = query.values.get(i);
            
            if (expectedType.equals("INT") || expectedType.equals("INTEGER")) {
                try {
                    Integer.parseInt(val);
                } catch (NumberFormatException e) {
                    return "Error: Value '" + val + "' is not a valid " + expectedType + " for column '" + schema.get(i).name + "'.";
                }
            }
        }

        if (storage.insertRow(query.tableName, query.values)) {
            return "1 row inserted into '" + query.tableName + "'.";
        }
        return "Error: Failed to insert row into '" + query.tableName + "'.";
    }

    private boolean evaluateCondition(String rowVal, String op, String condVal, String type) {
        if (type.equals("INT") || type.equals("INTEGER")) {
            try {
                int rVal = Integer.parseInt(rowVal);
                int cVal = Integer.parseInt(condVal);
                switch (op) {
                    case "=" -> { return rVal == cVal; }
                    case "<" -> { return rVal < cVal; }
                    case ">" -> { return rVal > cVal; }
                }
            } catch (NumberFormatException e) {
                return false;
            }
        } else {
            switch (op) {
                case "=" -> { return rowVal.equals(condVal); }
                case "<" -> { return rowVal.compareTo(condVal) < 0; }
                case ">" -> { return rowVal.compareTo(condVal) > 0; }
            }
        }
        return false;
    }

    private String executeSelect(SelectQuery query) {
        TableData data = storage.readTable(query.tableName);
        if (data == null) {
            return "Error: Failed to read table '" + query.tableName + "'.";
        }

        List<Integer> colIndices = new ArrayList<>();
        if (query.columns.size() == 1 && query.columns.get(0).equals("*")) {
            for (int i = 0; i < data.columns.size(); i++) {
                colIndices.add(i);
            }
        } else {
            for (String colName : query.columns) {
                int index = -1;
                for (int i = 0; i < data.columns.size(); i++) {
                    if (data.columns.get(i).name.equals(colName)) {
                        index = i;
                        break;
                    }
                }
                if (index != -1) {
                    colIndices.add(index);
                } else {
                    return "Error: Column '" + colName + "' not found.";
                }
            }
        }

        int whereColIdx = -1;
        String whereExpectedType = "STRING";
        if (query.hasWhere) {
            for (int i = 0; i < data.columns.size(); i++) {
                if (data.columns.get(i).name.equals(query.whereClause.column)) {
                    whereColIdx = i;
                    whereExpectedType = data.columns.get(i).type;
                    break;
                }
            }
            if (whereColIdx == -1) {
                return "Error: WHERE column '" + query.whereClause.column + "' not found.";
            }
            if (whereExpectedType.equals("INT") || whereExpectedType.equals("INTEGER")) {
                try {
                    Integer.parseInt(query.whereClause.value);
                } catch (NumberFormatException e) {
                    return "Error: WHERE clause value '" + query.whereClause.value + "' is not a valid " + whereExpectedType + ".";
                }
            }
        }

        StringBuilder sb = new StringBuilder();

        // Print header
        for (int i = 0; i < colIndices.size(); i++) {
            sb.append(data.columns.get(colIndices.get(i)).name);
            if (i < colIndices.size() - 1) sb.append(" | ");
        }
        sb.append("\n");
        sb.append("-".repeat(Math.max(10, colIndices.size() * 10))).append("\n");

        // Print rows
        int matchCount = 0;
        for (List<String> row : data.rows) {
            if (query.hasWhere && whereColIdx != -1) {
                if (!evaluateCondition(row.get(whereColIdx), query.whereClause.op, query.whereClause.value, whereExpectedType)) {
                    continue;
                }
            }
            
            for (int i = 0; i < colIndices.size(); i++) {
                sb.append(row.get(colIndices.get(i)));
                if (i < colIndices.size() - 1) sb.append(" | ");
            }
            sb.append("\n");
            matchCount++;
        }
        sb.append("(").append(matchCount).append(" rows)");
        return sb.toString();
    }
}
